pkg/streamingpromql/operators/functions/histograms.go

```diff
@@ -0,0 +1,410 @@
+// This means a group other than the one we are focused on may get completed first, but we
+// continue until the desired group is ready.
+func (h *HistogramFunctionOverInstantVector) accumulateUntilGroupComplete(ctx context.Context, g *bucketGroup) error {
+	for g.remainingSeriesCount > 0 {
+		s, err := h.inner.NextSeries(ctx)
+		if err != nil {
+			if errors.Is(err, types.EOS) {
+				return fmt.Errorf("exhausted series before all groups were completed: %w", err)
+			}
+			return err
+		}
+		thisSeriesGroups := h.seriesGroupPairs[h.currentInnerSeriesIndex]
+
+		// Native histograms only ever go to their original labelset.
+		// Floats only ever go to their group.
+		// It is possible that a series has both floats and histograms.
+		// It is also possible that both series groups are the same.
+		// The conflict in points is then detected in computeOutputSeriesForGroup.
+		h.saveNativeHistogramsToGroup(s.Histograms, thisSeriesGroups.nativeHistogramGroup)
+		err = h.saveFloatsToGroup(s.Floats, thisSeriesGroups.bucketValue, thisSeriesGroups.classicHistogramGroup)
+		if err != nil {
+			return err
+		}
+
+		// We are done with the fPoints, so return these now
+		// hPoint's are returned to the pool after computeOutputSeriesForGroup is finished with them as they may be copied to a group.
+		types.FPointSlicePool.Put(s.Floats, h.memoryConsumptionTracker)
+		h.currentInnerSeriesIndex++
+	}
+	return nil
+}
+
+// saveFloatsToGroup places each fPoint into a bucket with the upperBound set by the input series.
+func (h *HistogramFunctionOverInstantVector) saveFloatsToGroup(fPoints []promql.FPoint, le string, g *bucketGroup) error {
+	if len(fPoints) > 0 {
+		upperBound, err := strconv.ParseFloat(le, 64)
+		if err != nil {
+			// The le label was invalid. Record it:
+			h.annotations.Add(annotations.NewBadBucketLabelWarning(
+				h.innerSeriesMetricNames.GetMetricNameForSeries(h.currentInnerSeriesIndex),
+				le,
+				h.inner.ExpressionPosition(),
+			))
+		} else {
+			if g.pointBuckets == nil {
+				g.pointBuckets, err = pointBucketPool.Get(h.timeRange.StepCount, h.memoryConsumptionTracker)
+				if err != nil {
+					return err
+				}
+				g.pointBuckets = g.pointBuckets[:h.timeRange.StepCount]
+			}
+			for _, f := range fPoints {
+				pointIdx := h.timeRange.PointIndex(f.T)
+				g.pointBuckets[pointIdx] = append(
+					g.pointBuckets[pointIdx],
+					bucket{
+						upperBound: upperBound,
+						count:      f.F,
+					},
+				)
+			}
+		}
+	}
+	g.remainingSeriesCount--
+	return nil
+}
+
+// saveNativeHistogramsToGroup stores the given native histograms onto the given group.
+// There should only ever be one native histogram per a series or series group. In general,
+// they are per-series, but we handle them as parts of groups because they could hypothetically
+// have the same series name as a classic histogram.
+func (h *HistogramFunctionOverInstantVector) saveNativeHistogramsToGroup(hPoints []promql.HPoint, g *bucketGroup) {
+	if len(hPoints) > 0 {
```

Comment:
[nit] Similar comment to above - can we invert this `if` and return early if there are no histogram points?